### PR TITLE
Add Gradient Magnitude Similarity Deviation metric

### PR DESCRIPTION
**Pull request by nucli-vicky**
- [x] Add GMSD (Xue, Wufeng, et al. "Gradient magnitude similarity deviation: A highly efficient perceptual image quality index." IEEE transactions on image processing 23.2 (2013): 684-695)
- [x] Add tests for specific GMSD expected behavior with dummy inputs

### Checks to be done before submitting your PR

- [x] `python3 -m pytest deepinv/tests` runs successfully.
- [x] `black .` and `ruff check .` run successfully.
- [x] `make html` runs successfully (in the `docs/` directory).
- [x] Updated docstrings related to the changes (as applicable).
- [x] Added an entry to the [changelog.rst](https://github.com/deepinv/deepinv/blob/main/docs/source/changelog.rst).

### CODE REVIEW

File: deepinv/loss/metric/distortion.py [hunk 1]
```diff
@@ -990,3 +992,95 @@ def metric(self, x_net: Tensor, x: Tensor, *args, **kwargs) -> Tensor:
 
     def invert_metric(self, m: Tensor) -> Tensor:
         return 1.0 - m
+
+
+class GMSD(Metric):
+    r"""
+    Gradient Magnitude Similarity Deviation (GMSD) metric.
+
+    Calculates :math:`\text{GMSD}(x_net,x)`, as proposed in :footcite:t:`xue2013gradient`.
```
Comment (mh-nguyen712) [inline deepinv/loss/metric/distortion.py]:
```suggestion
    Calculates :math:`\text{GMSD}(x_{\mathrm{net}},x)`, as proposed in :footcite:t:`xue2013gradient`.
```

Comment (mh-nguyen712) [reply deepinv/loss/metric/distortion.py]: or even \hat{x} for notation consistency

File: deepinv/loss/metric/distortion.py [hunk 2]
```diff
@@ -990,3 +992,95 @@ def metric(self, x_net: Tensor, x: Tensor, *args, **kwargs) -> Tensor:
...
+    Prewitt operators :math:`h_x, h_y`,
+
+    .. math::
+        m_x = \sqrt{(x \ast h_x)^2 + (x \ast h_y)^2},\qquad
+        m_{\hat{x}} = \sqrt{(\hat{x} \ast h_x)^2 + (\hat{x} \ast h_y)^2},
+
+    where :math:`\ast` denotes 2D convolution. The pixel-wise GMS map is
+
+    .. math::
+        \text{GMS}(\hat{x},x) = \frac{2 m_x m_{\hat{x}} + c}{m_x^2 + m_{\hat{x}}^2 + c},
```
Comment (mh-nguyen712) [inline deepinv/loss/metric/distortion.py:1015]: precise what is c?

File: deepinv/loss/metric/distortion.py [hunk 3]
```diff
@@ -990,3 +992,95 @@ def metric(self, x_net: Tensor, x: Tensor, *args, **kwargs) -> Tensor:
...
+            )
+        hx = torch.tensor(
+            [[1 / 3, 0, -1 / 3], [1 / 3, 0, -1 / 3], [1 / 3, 0, -1 / 3]],
+            device=x_net.device,
+        ).view(1, 1, 3, 3)
+
+        hy = torch.tensor(
+            [[1 / 3, 1 / 3, 1 / 3], [0, 0, 0], [-1 / 3, -1 / 3, -1 / 3]],
+            device=x_net.device,
+        ).view(1, 1, 3, 3)
```
Comment (mh-nguyen712) [inline deepinv/loss/metric/distortion.py]: Move this to __init__

File: deepinv/loss/metric/distortion.py [hunk 4]
```diff
@@ -990,3 +992,95 @@ def metric(self, x_net: Tensor, x: Tensor, *args, **kwargs) -> Tensor:
...
+
+        grad_mag_x = torch.sqrt(
+            conv2d(x, hx, padding="replicate") ** 2
+            + conv2d(x, hy, padding="replicate") ** 2
+        )
```
Comment (mh-nguyen712) [inline deepinv/loss/metric/distortion.py]: there is torch.hypot which does the same thing in a single kernel call
```suggestion
        grad_mag_x = torch.hypot(
            conv2d(x, hx, padding="replicate"),
            conv2d(x, hy, padding="replicate")
        )
```

File: deepinv/loss/metric/distortion.py [hunk 5]
```diff
@@ -990,3 +992,95 @@ def metric(self, x_net: Tensor, x: Tensor, *args, **kwargs) -> Tensor:
...
+    def metric(self, x_net: Tensor, x: Tensor, *args, **kwargs) -> Tensor:
+        if x_net.shape != x.shape:
+            raise ValueError(
+                f"x_net and x must be same shape, but got {tuple(x_net.shape)} and {tuple(x.shape)}"
+            )
+        if x_net.device != x.device:  # pragma: no cover
+            raise ValueError(
+                f"GMSD requires x_net {x_net.device} and x {x.device} to be on the same device, "
+            )
+        if len(x_net.shape) != 4 or x_net.shape[1] != 1:
```
Comment (mh-nguyen712) [inline deepinv/loss/metric/distortion.py]: Can we extend to RGB images? Eg by taking mean over channel dimension?

File: deepinv/loss/metric/distortion.py [hunk 6]
```diff
@@ -990,3 +992,95 @@ def metric(self, x_net: Tensor, x: Tensor, *args, **kwargs) -> Tensor:
...
+        If `tuple` of `int`, cropping is performed over the last `len(center_crop)` dimensions. If positive values are provided, a standard center crop is applied.
+        If negative (or zero) values are passed, cropping will be done by removing `center_crop` pixels from the borders (useful when tensors vary in size across the dataset).
+    """
+
+    def __init__(self, c: float = 0.0026, **kwargs):
+        super().__init__(**kwargs)
+        self.c = c
+
+    def metric(self, x_net: Tensor, x: Tensor, *args, **kwargs) -> Tensor:
+        if x_net.shape != x.shape:
```
Comment (mh-nguyen712) [inline deepinv/loss/metric/distortion.py:1059]: Can we support broadcasting here, in the batch dimension?

**Review by mh-nguyen712 — CHANGES_REQUESTED** [review]
Many thanks for this addition, looks great!

File: deepinv/loss/metric/distortion.py (hunk 5 above)
Comment (nucli-vicky) [reply deepinv/loss/metric/distortion.py]: So calculating GMSD over all channels, and taking the mean over those GMSD scores? Can do!

File: deepinv/loss/metric/distortion.py (hunk 3 above)
Comment (nucli-vicky) [reply deepinv/loss/metric/distortion.py]: I moved it to `_build_hx_hy`, as this allows the metric to handle inputs with both cpu and cuda devices.
When initializing `hx`, `hy` at init, calling `.to(device)` would probably be slower than constructing it on the proper device each time? Not sure, correct me if I'm wrong here :)

File: deepinv/loss/metric/distortion.py (hunk 4 above)
Comment (nucli-vicky) [reply deepinv/loss/metric/distortion.py]: cool, thanks!

File: deepinv/loss/metric/distortion.py (hunk 3 above)
Comment (mh-nguyen712) [reply deepinv/loss/metric/distortion.py]: Yes you're right, calling `to(device)` is slower.

File: deepinv/loss/metric/distortion.py (hunk 5 above)
Comment (mh-nguyen712) [reply deepinv/loss/metric/distortion.py]: Yes I think we can compute the mean of scores over channels, just need to be clear in the docs

File: deepinv/loss/metric/distortion.py (hunk 6 above)
Comment (nucli-vicky) [reply deepinv/loss/metric/distortion.py:1059]: not sure if this is ideal, are there any scenario's where metrics would be used with one of the inputs `B=1` and other input not?

Would this happen for `x_net`, `x`, or both possible?

File: deepinv/loss/metric/distortion.py (hunk 2 above)
Comment (nucli-vicky) [reply deepinv/loss/metric/distortion.py:1015]: done

File: deepinv/loss/metric/distortion.py (hunk 5 above)
Comment (nucli-vicky) [reply deepinv/loss/metric/distortion.py]: Done

File: deepinv/loss/metric/distortion.py (hunk 6 above)
Comment (mh-nguyen712) [reply deepinv/loss/metric/distortion.py:1059]: Hmm agree!

**Comment by nucli-vicky** [comment]
@mh-nguyen712  thanks for the comments! I tried to address them here.

**Review by mh-nguyen712 — APPROVED** [review]
